### PR TITLE
Dictionary endpoint

### PR DESCRIPTION
**Pull request by richwhitjr**
Sometimes the features stored in TFExamples doesn't map well to a Pandas Dataframe.  For example sparse features or even list of floats for one key.  This PR adds a middle layer that
first reads the data as dictionaries then passes it through to the pandas dataframe.  If a user
has a custom need though like above they can use the dictionary endpoint directly.

### CODE REVIEW

File: spotify_tensorflow/dictionary_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+class DictonaryEndpoint(object):
+    @classmethod
+    def read_dataset(cls, dataset_path, take=sys.maxsize, feature_mapping_fn=None):
+        """
+        Read a TF dataset and load it into a Dictionary of Numpy Arrays.
+
+        :param dataset_path: Path to the TF Records Dataset
+        :param take: Number of records to read
+        :param unpack_multispec: Returns an array of DataFrames, order is the same as the one used
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: Where is this used?

File: spotify_tensorflow/dictionary_endpoint.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        return six.next(cls.batch_iterator(dataset_path, take, feature_mapping_fn))
+
+    @classmethod
+    def batch_iterator(cls, dataset_path, batch_size=10000, feature_mapping_fn=None):
+        """
+        Read a TF dataset in batches, each one yielded as a Dictonary.
+
+        :param dataset_path: Path to the TF Records Dataset
+        :param batch_size: Size of each batches
+        :param unpack_multispec: Returns an array of DataFrames, order is the same as the one used
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: Where is this (`unpack_multispec`) used?

File: spotify_tensorflow/dictionary_endpoint.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        when building the MultiFeatureSpec in Featran
+        :param feature_mapping_fn: Override the TF record reading function
+        :return: A Python Generator, yielding batches of data in a Dictonary
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: -> `Dictionary`

File: spotify_tensorflow/dictionary_endpoint.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    def read_dataset(cls, dataset_path, take=sys.maxsize, feature_mapping_fn=None):
+        """
+        Read a TF dataset and load it into a Dictionary of Numpy Arrays.
+
+        :param dataset_path: Path to the TF Records Dataset
+        :param take: Number of records to read
+        :param unpack_multispec: Returns an array of DataFrames, order is the same as the one used
+        when building the MultiFeatureSpec in Featran
+        :param feature_mapping_fn: Override the TF record reading function
+        :return: A Dictonary containing the dataset
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: -> `Dictionary`

File: spotify_tensorflow/dictionary_endpoint.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+from .dataset import Datasets
+
+
+FLAGS = tf.flags.FLAGS
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class DictonaryEndpoint(object):
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: -> `DictionaryEndpoint`

Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: What would you think of instead of exposing these:
 * `DictionaryEndpoint`
 * `DataFrameEndpoint`

as separate classes, instead expose it via `Datasets` API, so it would be something like:

```
Datasets.dict.read_dataset(...)
Datasets.dict.batch_iterator(...)
```

and for dataframe:

```
Datasets.dataframe.read_dataset(...)
Datasets.dataframe.batch_iterator(...)
```

and we would mark both `DictionaryEndpoint` and `DataFrameEndpoint` as private. This should simplify the API i believe.

What do you think? also pinging @yonromai

File: spotify_tensorflow/dictionary_endpoint.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+    class __FeatureGenerator(object):
+        def __init__(self, training_it, batch_size, context):
+            self.batch_size = batch_size
+            self.batch_iter = training_it.get_next()
+            self.context = context
+            self.buff = None
+
+        def __iter__(self):
+            logger.info("Starting TF Session...")
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: Whenever I see `Starting ... ` comments I like when there is `Closing ...` or `Ending ...` (sometimes with duration) this simplifies debugging. What do you think?

Comment (ravwojdyla) [reply spotify_tensorflow/dictionary_endpoint.py]: Actually I guess, there is `End of dataset`, wonder if we should add duration of the session?

File: spotify_tensorflow/dictionary_endpoint.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+                for k in self.context.features.keys():
+                    self.buff[k] = first_result[k]
+            while self._get_buff_size() < self.batch_size:
+                t = timeit.default_timer()
+                current_batch = sess.run(self.batch_iter)
+                for k in self.context.features.keys():
+                    self.buff[k] = self._append(self.buff[k], current_batch[k])
+                logger.info("Fetched %d / %s records (%4d TFExamples/s)" % (
+                    self._get_buff_size(),
+                    str(self.batch_size) if self.batch_size < sys.maxsize else "?",
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: Why is this necessary? Should this be asserted somewhere?

File: spotify_tensorflow/dictionary_endpoint.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                    FLAGS.batch_size / (timeit.default_timer() - t)))
+            ret = OrderedDict()
+            for k in list(self.context.features.keys()):
+                ret[k] = self.buff[k][:self.batch_size]
+                self.buff[k] = self.buff[k][self.batch_size:]
```
Comment (ravwojdyla) [inline spotify_tensorflow/dictionary_endpoint.py]: What is this line for?

File: spotify_tensorflow/dictionary_endpoint.py (hunk 7 above)
Comment (richwhitjr) [reply spotify_tensorflow/dictionary_endpoint.py]: Not sure some of this code was refactored from @yonromai.  Any idea?

File: spotify_tensorflow/dictionary_endpoint.py (hunk 8 above)
Comment (richwhitjr) [reply spotify_tensorflow/dictionary_endpoint.py]: This is to segment the currently buffered data into the values to return and to keep the rest stored on the buff object.  Because each read can be a different size than the batch size.

File: spotify_tensorflow/dictionary_endpoint.py (hunk 5 above)
Comment (richwhitjr) [reply spotify_tensorflow/dictionary_endpoint.py]: Refactored in this way.  Let me know what you think.
